homeassistant/components/remote/xiaomi_miio.py

```diff
@@ -0,0 +1,261 @@
+
+    # Create handler
+    _LOGGER.info("Initializing with host %s (token %s...)", host, token[:5])
+    device = ChuangmiIr(host, token)
+
+    try:
+        device.info()
+
+    # This should be DeviceError but python-miio returns wrong except.
+    except ChecksumError as ex:
+        _LOGGER.error("Token not accepted by device : %s", ex)
+        return
+
+    if PLATFORM not in hass.data:
+        hass.data[PLATFORM] = {}
+
+    entity_id = config.get(CONF_NAME, "xiaomi_miio_" +
+                           host.replace('.', '_'))
+    slot = config.get(CONF_SLOT)
+    timeout = config.get(CONF_TIMEOUT)
+
+    hidden = config.get(ATTR_HIDDEN)
+
+    xiaomi_miio_remote = XiaomiMiioRemote(
+        entity_id, device, slot, timeout, hidden, config.get(CONF_COMMANDS))
+
+    hass.data[PLATFORM][host] = xiaomi_miio_remote
+
+    async_add_devices([xiaomi_miio_remote])
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Handle a learn command."""
+        if service.service != SERVICE_LEARN:
+            _LOGGER.error("We should not handle service: %s", service.service)
+            return
+
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+        entity = None
+        for remote in hass.data[PLATFORM].values():
+            if remote.entity_id == entity_id:
+                entity = remote
+
+        if not entity:
+            _LOGGER.error("entity_id: '%s' not found", entity_id)
+            return
+
+        device = entity.device
+
+        slot = service.data.get(CONF_SLOT, entity.slot)
+
+        yield from hass.async_add_job(device.learn, slot)
+
+        timeout = service.data.get(CONF_TIMEOUT, entity.timeout)
+
+        _LOGGER.info("Press the key you want Home Assistant to learn")
+        start_time = utcnow()
+        while (utcnow() - start_time) < timedelta(seconds=timeout):
+            message = yield from hass.async_add_job(
+                device.read, slot)
+            _LOGGER.debug("Message recieved from device: '%s'", message)
+
+            if 'code' in message and message['code']:
+                log_msg = "Received command is: {}".format(message['code'])
+                _LOGGER.info(log_msg)
+                hass.components.persistent_notification.async_create(
+                    log_msg, title='Xiaomi Miio Remote')
+                return
+
+            if ('error' in message and
+                    message['error']['message'] == "learn timeout"):
+                yield from hass.async_add_job(device.learn, slot)
+
+            yield from asyncio.sleep(1, loop=hass.loop)
+
+        _LOGGER.error("Timeout. No infrared command captured")
+        hass.components.persistent_notification.async_create(
+            "Timeout. No infrared command captured",
+            title='Xiaomi Miio Remote')
+
+    hass.services.async_register(DOMAIN, SERVICE_LEARN, async_service_handler,
+                                 schema=LEARN_COMMAND_SCHEMA)
+
+
+class XiaomiMiioRemote(Entity):
```

Comment:
You need impolement this interface: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/remote/__init__.py#L177